### PR TITLE
Fix for relative markdown image paths

### LINKED ISSUE #3069: Markdown relative paths for images not rendering
Issue Type: <b>Bug</b>

There is an issue where we are not honoring relative paths in markdowns. We need to fix this. 

I verified that the absolute file paths will work.

Azure Data Studio version: Azure Data Studio 1.2.1-insider (f7d92caae51a679934a9f945ea0b48e89f02e1de, 2018-10-23T00:30:42.493Z)
OS version: Darwin x64 17.7.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz (8 x 2200)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|16.00GB (0.09GB free)|
|Process Argv|/private/var/folders/vc/2qtbnhx57m56_lzr6h8gv8gh0000gn/T/AppTranslocation/3EC7DA19-0CFB-44EB-8B97-3961178BD401/d/Azure Data Studio.app/Contents/MacOS/Electron|
|Screen Reader|no|
|VM|0%|

</details>

### PR DESCRIPTION
**Pull request by chlafreniere**
Fixes #3069. Note that we only allow images to be loaded from either the same folder as a notebook, or a subfolder of where the notebook is saved.

I wasn't familiar with the concept of vscode-resource before this change, so the following link may be helpful to others: https://code.visualstudio.com/api/extension-guides/webview.

If notebooks are unsaved, relative images will not load, as they do not have an fsPath associated with them. If anyone has any cool ideas for how to solve that (does it need solving? Maybe we can doc it?), more than happy to entertain any ideas.

@kevcunnane mentioned that we may want to allow any images saved in any of the workspace folders. More than happy to create an issue to track that, as I have a feeling that the April release will be closing fairly soon 😄.

### CODE REVIEW

File: src/sql/parts/notebook/cellViews/textCell.component.ts [hunk 1]
```diff
@@ -133,8 +134,9 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 			} else {
 				this._content = this.sanitizeContent(this.cellModel.source);
 			}
-			// todo: pass in the notebook filename instead of undefined value
-			this._commandService.executeCommand<string>('notebook.showPreview', undefined, this._content).then((htmlcontent) => {
+
+			this._commandService.executeCommand<string>('notebook.showPreview', this.cellModel.notebookModel.notebookUri, this._content).then((htmlcontent) => {
+				htmlcontent = this.convertVsCodeResourceToFileInSubDirectories(htmlcontent);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/cellViews/textCell.component.ts]: super nitpicky: I think it's `Vscode` in most places not `VsCode`.

File: src/sql/parts/notebook/cellViews/textCell.component.ts [hunk 2]
```diff
@@ -149,6 +151,24 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 		return content;
 	}
 
+	// Only replace vscode-resource with file when in the same (or a sub) directory
+	// This matches Jupyter Notebook viewer behavior
+	private convertVsCodeResourceToFileInSubDirectories(htmlContent: string): string {
+		let htmlContentCopy = htmlContent;
+		while (htmlContentCopy.search('(?<=img src=\"vscode-resource:)') > 0) {
+			let pathStartIndex = htmlContentCopy.search('(?<=img src=\"vscode-resource:)');
+			let pathEndIndex = htmlContentCopy.slice(pathStartIndex).indexOf('\" ') + pathStartIndex;
```
Comment (kevcunnane) [inline src/sql/parts/notebook/cellViews/textCell.component.ts]: Questions:
- Could there be a case where someone escapes this as part of the path, e.g. `\\" ` and this matches incorrect? Am I just being paranoid?
- Slicing large text could be super inefficient? Looking at [w3 docs](https://www.w3schools.com/jsref/jsref_indexof.asp), I think you can pass in the start position to search from and avoid this issue. Can you do that instead? It should be a lot more efficient.
```
			let pathEndIndex = htmlContentCopy.indexOf('\" ', pathStartIndex);
```

File: src/sql/parts/notebook/cellViews/textCell.component.ts [hunk 3]
```diff
@@ -149,6 +151,24 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
...
+			let filePath = htmlContentCopy.substring(pathStartIndex, pathEndIndex);
+			// If the asset is in the same folder or a subfolder, replace 'vscode-resource:' with 'file:', so the image is visible
+			if (!path.relative(path.dirname(this.cellModel.notebookModel.notebookUri.fsPath), filePath).includes('..')) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/cellViews/textCell.component.ts]: For relative paths how will this `path.relative` end up looking? For example:
```
./its/a/subdir/with/../some/../in/the/name
../actualRelativeDir
```

Comment (chlafreniere) [reply src/sql/parts/notebook/cellViews/textCell.component.ts]: ../actualRelativeDir

File: src/sql/parts/notebook/cellViews/textCell.component.ts (hunk 2 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/cellViews/textCell.component.ts]: Addressed the second bullet. For the first one, someone would have to have a folder/file name that is literally called folderAbove\" blah"

i.e. the folder would have to start with a quote, followed by a space.

I'm not convinced this is a mainline scenario, even if it has a tiny chance at being a bug.

Comment (kevcunnane) [reply src/sql/parts/notebook/cellViews/textCell.component.ts]: Sounds good, thanks

**Pull request merged**